### PR TITLE
HPU Adaptations for PyTorch TinyImagenet Interactive API

### PR DESCRIPTION
**Pull request by Supriya-Krishnamurthi**
Refer README.md for detailed instructions

### CODE REVIEW

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# PyTorch_TinyImageNet
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md]: General: Ensure consistency across your markdown for sections, subsections, lists, code snippets etc.
Use [this](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) as a good starting point.

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+## **How to run this tutorial (without TLC and locally as a simulation):**
+<br/>
+
+### 0. If you haven't done so already, create a virtual environment(Python>=3.8), install OpenFL, and upgrade pip:
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md]: As I understand, Habana stack currently doesn't work well inside `venv`. Is this intended?

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+```
+
+2. Target the Gaudi HPU device:
+
+```
+device = torch.device("hpu")
+```
+3. Move the model to the device:
+
+**This step should be executed before initializing any optimizers. In the upcoming versions the order dependency will be eliminated**
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md]: Habana backend has 2 specific workarounds, as understood from the document.
Suggest continuing the usual steps for PyTorch/Tensorflow, and move the 2 "workarounds" under a separate subsection calling out "Sharp bits" or "Known Issues"
Also worth mentioning if any of them do not apply to Tensorflow (I'm not sure)

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+    "from openfl.interface.interactive_api.experiment import TaskInterface, DataInterface, ModelInterface, FLExperiment\n",
+    "from copy import deepcopy\n",
+    "import torchvision\n",
+    "from torchvision import transforms as T\n",
+    "from torch.utils.data import Dataset\n",
+    "from torch.utils.data import DataLoader\n",
+    "import tqdm\n",
+    "\n",
+    "import habana_frameworks.torch.core as htcore\n",
+    "import habana_frameworks.torch.hpu as hthpu\n",
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb]: This is an unused import. Can be removed.

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "79021778",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_net.to('hpu')\n",
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb:313]: Leave a `# NOTE: This is a workaround for ...` here briefing the issue.

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+    "\n",
+    "    for data, target in train_loader:\n",
+    "        data, target = torch.tensor(data).to(device), torch.tensor(\n",
+    "            target).to(device)\n",
+    "        optimizer.zero_grad()\n",
+    "        output = net_model(data)\n",
+    "        loss = loss_fn(output=output, target=target)\n",
+    "        loss.backward()\n",
+    "        # API call to trigger execution\n",
+    "        htcore.mark_step()\n",
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb:388]: Same as previous comment for `NOTE`

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,221 @@
+absl-py==1.2.0
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/requirements.txt]: This file can be removed if you have validated.

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "895288d0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#!pip install -r requirements.txt"
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb]: This can be removed if you have validated without additional packages. To avoid confusion to the user.

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/envoy/requirements.txt [hunk 9]
```diff
@@ -0,0 +1 @@
+Pillow==9.0.1
```
Comment (MasterSkepticista) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/envoy/requirements.txt:1]: This file can be removed if you have validated.

**Review by MasterSkepticista — CHANGES_REQUESTED** [review]

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+- A Jupyter Server URL will appear in your terminal. In your browser, proceed to that link. Once the webpage loads, click on the pytorch_tinyimagenet.ipynb file. 
+- To run the experiment, select the icon that looks like two triangles to "Restart Kernel and Run All Cells". 
+- You will notice activity in your terminals as the experiment runs, and when the experiment is finished the director terminal will display a message that the experiment has finished successfully.  
+
+<br/>
+
+### HPU Tutorials
+#### The name of the file/example that contain HPU adaptations start with "HPU". For example:
+
+ HPU_pytorch_tinyimagenet.ipynb placed under workspace folder of this example is similar to pytorch_tinyimagenet.ipynb but contains HPU adaptations and necessary packages to enable smooth execution on Habana accelerators (Gaudi)
```
Comment (sun1lach) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md]: remove the comparison part.

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+ All the above execution steps remain same for HPU examples but as pre-requisite it needs some additional environment setup and Habana supported package installations which is explained below.
+
+ #### Environment Setup
```
Comment (sun1lach) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md]: Provide AWS DL1 setup details and tested environment.

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+ To perform an installation of the full driver and SynapseAI software stack independently (as an on-premise installation), you can refer to the dedicated Setup and Install [GitHub repository](https://github.com/HabanaAI/Setup_and_Install) for detailed instructions. Make sure to install the PyTorch and Tensorflow packages provided by Habana. These packages are optimized for Habana Gaudi HPU. Installing public PyTorch and Tensorflow packages is not supported. Refer the same above mentioned github repository for Pytorch and Tensorflow package installations.
```
Comment (sun1lach) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md]: No on-premise mentions.

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+ #### HPU Adaptations For Tensorflow Examples
```
Comment (sun1lach) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md]: Extract the HPU info out of the example folder.

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md (hunk 1 above)
Comment (Supriya-Krishnamurthi) [reply openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md]: Updated

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md (hunk 2 above)
Comment (Supriya-Krishnamurthi) [reply openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md]: Latest installation scripts it supports venv

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md (hunk 3 above)
Comment (Supriya-Krishnamurthi) [reply openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/README.md]: There is only one workaround to move model to HPU before initializing any optimizers. Like discussed I have mentioned it as workaround in the statement

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb (hunk 6 above)
Comment (Supriya-Krishnamurthi) [reply openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/workspace/hpu_pytorch_tinyimagenet.ipynb:388]: This is not a workaround, it is part of the HPU adaptations steps as mentioned in README.md

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/envoy/requirements.txt (hunk 9 above)
Comment (Supriya-Krishnamurthi) [reply openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/envoy/requirements.txt:1]: This requirement.txt file in envoy folder is needed

**Review by MasterSkepticista — APPROVED** [review]

**Review by sun1lach — APPROVED** [review]

File: openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/envoy/envoy_config_1.yaml [hunk 14]
```diff
@@ -0,0 +1,10 @@
+params:
+  cuda_devices: []
+
+optional_plugin_components: {}
+
+shard_descriptor:
+  template: tinyimagenet_shard_descriptor.TinyImageNetShardDescriptor
+  params:
+    data_folder: tinyimagenet_data
+    rank_worldsize: 1,2
```
Comment (shreyasnc1) [inline openfl-tutorials/interactive_api/HPU/PyTorch_TinyImageNet/envoy/envoy_config_1.yaml]: Should this be 1,1 (considering we had added only envoy)

**Review by shreyasnc1 — APPROVED** [review]

**Comment by psfoley** [comment]
Related to https://github.com/intel/openfl/issues/642
